File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
def solution(queue1, queue2):
    '''
    길이가 같은 두 개의 큐에서 하나의 큐에서 빼서 다른 큐에 원소를 넣는다
    1번의 POP과 1번의 INSERT가 작업 1회
    큐이므로 먼저 집어넣은 원소를 빼는 것
    
    그래서 각 큐의 원소의 합을 같도록 하기
    리턴 : 최소 연산 횟수
    
    최적의 해니까 BFS 같다 
    
    1에서 먼저? 2에서 먼저 둘중에서 최소
    
    첫 번째에서 시작할 때 합이 절반보다 작으면 ? 2번째 
    
    while queue1 :
            if start_dif == 0 :
                # break
                return cnt
            if sum(queue1) < sum(queue2) :
                pick = queue2.popleft()
                queue1.append(pick)
                cnt+=1
            elif sum(queue1) > sum(queue2) :
                pick = queue1.popleft()
                queue2.append(pick)
                cnt+=1
                
        return -1
    '''
    from collections import deque
    queue1, queue2 = deque(queue1), deque(queue2)
    
    def bfs1(queue1, queue2) :
        cnt = 0
        if sum(queue1) < sum(queue2) :
            queue1 , queue2 = queue2, queue1
        start_dif = sum(queue1)-sum(queue2)
        
        while queue1 :
            if cnt > 4* len(queue1):
                return -1
            if start_dif == 0 :
                # break
                return cnt
            if  start_dif < 0 :
                pick = queue2.popleft()
                queue1.append(pick)
                start_dif += 2*pick
                cnt+=1
            elif start_dif > 0 :
                pick = queue1.popleft()
                queue2.append(pick)
                start_dif -= 2*pick
                cnt+=1
                
        return -1
    
    
    return bfs1(queue1,queue2)

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;
import java.io.*;
import java.util.stream.*;
class Solution {
    /*
    1.queue1과 queue2의 sum을 비교한다.
    2.작은 쪽에서 큰 쪽으로 수를 넘긴다.
    3. cnt++ 한다.
    
    다시 1번..
    
    
    */
    public int solution(int[] queue1, int[] queue2) {
        List<Integer> list1 = Arrays.stream(queue1)
            .boxed()
            .collect(Collectors.toList());
        List<Integer> list2 = Arrays.stream(queue2)
            .boxed()
            .collect(Collectors.toList());
        Deque<Integer> deque1 = new ArrayDeque<>(list1);
        Deque<Integer> deque2 = new ArrayDeque<>(list2);
        
        long sum1 = 0L;
        long sum2 = 0L;
        sum1 = list1.stream()
            .mapToInt(Integer::intValue)
            .sum();
        sum2 = list2.stream()
            .mapToInt(Integer::intValue)
            .sum();
        
        int size = 4 * queue1.length;
        
        
        if (sum1 == sum2 ){return 0;}
        int num1 = 0;
        int num2 = 0;
        int cnt = 0 ;
        
        while(cnt <= size && !deque1.isEmpty() && !deque2.isEmpty()){
            num1 = deque1.peekFirst();
            num2 = deque2.peekFirst();
            
            if(sum1>sum2){
                deque2.addLast(deque1.removeFirst());
                sum1-=num1;
                sum2+=num1;
            }else{
                deque1.addLast(deque2.removeFirst());
                sum1+=num2;
                sum2-=num2;
            }
            cnt++;
            if(sum1==sum2){
                return cnt;
            }
            
        }
        return -1;
    }
}